### PR TITLE
feat(CNLS/tools): Implement basic error/exception system

### PR DESCRIPTION
**Pull request by Fangop**
Here is a draft for exception system.

Sometimes, we comes to a situation that should stop the process and inform the users are defined as exceptions in python.
[Built-in Exceptions](https://docs.python.org/3/library/exceptions.html)

The error is included in exception.
The pyStoNED brings in at least 2 types of exceptions here.

### Additive CNLS with CRS
The additive CNLS model with CRS does not exist (or be needed).
Hence when creating and additive CNLS model with CRS should raise an exception, since it is not an error but an exception of existing model.

The following codes may halt and bring out an exception
```python
# import packages
from pystoned import CNLS
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_CRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = CET_ADDI, fun = FUN_PROD, rts = RTS_CRS)
```
Please help me justify the discussion above and polish the exception message.

### Retrieving variables without optimization
User should optimize the model before retrieving and printing any variables.
If not, the program will halt the program and inform the users to optimize the model.

The following codes may halt and bring out an exception
```python
# import packages
from pystoned import CNLS
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = CET_ADDI, fun = FUN_PROD, rts = RTS_VRS)


model.display_alpha()
```

### Construct model with unknown parameters

User should construct a model with constant labels in `pystoned.constant`.
If a random string is giving, the program will halt the program and inform the users the model parameter is not defined.

This example construct a model with a random string as cet, causing the value error.
```python
from pystoned import CNLS
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = "Not an CET label", fun = FUN_PROD, rts = RTS_VRS)
```

Note: This does not affect the default setting.

### Invalid email address
When users using remote optimization, the user may use incorrect string(not an email address and OPT_LOCAL label).

This should leads to a halt and informs the user.
```python
# import packages
from pystoned import CNLS
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = CET_ADDI, fun = FUN_PROD, rts = RTS_VRS)

model.optimize(email="NotAnEmailAddress")
```

### Optimization multiplicative model without specifying solvers.
When users using local optimization, the user should specify the solver for optimization.

This should leads to a halt and informs the user to choose a installed solver.
```python
# import packages
from pystoned import CNLS
from pystoned.constant import CET_MULT, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = CET_MULT, fun = FUN_PROD, rts = RTS_VRS)

model.optimize(email=OPT_LOCAL)
```

These modification is a draft for discussing error/exception types, situation, and the messages.
Hence only the CNLS and the utils/tools module are modified as examples.
Any other exceptions and errors can be included and discussed! 

Thanks!

Note: This part may not be included in the document. Since the document should indicate the right way to use the program, and here is for prevention of wrong ways.

### CODE REVIEW

File: pystoned/CNLS.py [hunk 1]
```diff
@@ -135,8 +135,7 @@ def regression_rule(model, i):
 
                 return regression_rule
             elif self.rts == RTS_CRS:
-                # TODO(warning handling): replace with model requested not exist
-                return False
+                raise Exception("The Additive CNLS with constant return to scale does not exist")
```
Comment (ds2010) [inline pystoned/CNLS.py:138]: Although additive CNLS-CRS models are rarely used in the application, they should be added to our package. So, could you please add CRS specifications to all additive CNLS models?

File: pystoned/CNLS.py [hunk 2]
```diff
@@ -173,8 +171,7 @@ def log_rule(model, i):
 
                 return log_rule
 
-        # TODO(error handling): replace with undefined model attribute
-        return False
+        raise ValueError("Undefined model parameter")
```
Comment (ds2010) [inline pystoned/CNLS.py:174]: `raise ValueError("Undefined model parameters")`  ???

File: pystoned/CNLS.py [hunk 3]
```diff
@@ -291,28 +277,23 @@ def get_beta(self):
...
         """Return beta value by array"""
         if self.optimization_status == 0:
-            print("Model isn't optimized. Use optimize() method to estimate the model.")
-            return False
+            raise Exception("Model isn't optimized. Use optimize() method to estimate the model.")
         if type(self.z) == type(None):
-            # TODO: Replace print by warning
-            print("Without z variable")
-            return
+            raise Exception("Cannot retrieve z variables, since the model is built without z variables.")
```
Comment (ds2010) [inline pystoned/CNLS.py:289]: should be:

` raise Exception("Estimated coefficient (lambda) cannot be retrieved due to no z variable included in the model.")`

File: pystoned/CNLS.py [hunk 4]
```diff
@@ -197,8 +194,7 @@ def afriat_rule(model, i, h):
 
                 return afriat_rule
             elif self.rts == RTS_CRS:
-                # TODO(warning handling): replace with model requested not exist
-                return False
+                raise Exception("The Additive CNLS with constant return to scale does not exist")
```
Comment (ds2010) [inline pystoned/CNLS.py:197]: Please add CRS specifications to all additive CNLS models

**Review by ds2010 — CHANGES_REQUESTED** [review]
Here are the comments.

File: pystoned/CNLS.py [hunk 5]
```diff
@@ -223,46 +219,38 @@ def afriat_rule(model, i, h):
...
         """Display lamda value"""
         if self.optimization_status == 0:
-            print("Model isn't optimized. Use optimize() method to estimate the model.")
-            return False
+            raise Exception("Model isn't optimized. Use optimize() method to estimate the model.")
         if type(self.z) == type(None):
-            # TODO: Replace print by warning
-            print("Without z variable")
-            return
+            raise Exception("Cannot retrieve z variables, since the model is built without z variables.")
```
Comment (ds2010) [inline pystoned/CNLS.py:247]: should be:

` raise Exception("Estimated coefficient (lambda) cannot be retrieved due to no z variable included in the model.")`

**Review by ds2010 — CHANGES_REQUESTED** [review]

**Comment by ds2010** [comment]
Thanks for your work. Another verification before estimating the model could be added. The length of `x` , `y`, and `z` should be equal. The pre-specified `tau` should be in `(0, 1)`. Here is an example in R package: 

https://github.com/cran/frontiles/blob/master/R/alphascore.r#L8

File: pystoned/CNLS.py (hunk 2 above)
Comment (Fangop) [reply pystoned/CNLS.py]: Resolved!

File: pystoned/CNLS.py (hunk 3 above)
Comment (Fangop) [reply pystoned/CNLS.py]: The message has been modified.

File: pystoned/CNLS.py (hunk 4 above)
Comment (Fangop) [reply pystoned/CNLS.py]: The CRS with additive model has been added to CNLS model. The other related model will be modified in another fix commit.

File: pystoned/CNLS.py (hunk 5 above)
Comment (Fangop) [reply pystoned/CNLS.py]: The message has been modified.

**Comment by Fangop** [comment]
@ds2010 , it comes to me that the intercept term should not exist for CRS model.
Hence the exception will also occurs when retrieving alpha with a CRS model.
The exception has been added in the last fixed commit.


This will also raise an exception `Exception: Estimated intercept (alpha) cannot be retrieved due the CRS model is built without intercept.`
```python
from pystoned import CNLS
from pystoned.constant import FUN_PROD, OPT_LOCAL, RTS_CRS, CET_MULT
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y, x=data.x, z=None,
                    cet = CET_MULT, fun = FUN_PROD, rts = RTS_CRS)

model.optimize(email="test@site.com")

model.display_alpha()
```

File: pystoned/CNLS.py [hunk 6]
```diff
@@ -223,46 +239,40 @@ def afriat_rule(model, i, h):
...
         print(self.display_status)
 
     def display_alpha(self):
         """Display alpha value"""
         if self.optimization_status == 0:
-            print("Model isn't optimized. Use optimize() method to estimate the model.")
-            return False
+            raise Exception("Model isn't optimized. Use optimize() method to estimate the model.")
+        if self.rts == RTS_CRS:
+            raise Exception("Estimated intercept (alpha) cannot be retrieved due the CRS model is built without intercept.")
```
Comment (ds2010) [inline pystoned/CNLS.py:255]: could be:

`raise Exception("Estimated intercept (alpha) cannot be retrieved due to the constant returns-to-scale assumption.")`

**Review by ds2010 — CHANGES_REQUESTED** [review]
Here is a further revision.

**Review by ds2010 — APPROVED** [review]

File: pystoned/CNLS.py (hunk 6 above)
Comment (Fangop) [reply pystoned/CNLS.py]: Thanks for this precise revision!.

Comment (Fangop) [reply pystoned/CNLS.py]: The message has been updated

**Comment by Fangop** [comment]
If the structure of the error and exception handling is reasonable, we can merge this to the upstream pyStoNED.
And this also revealed some works for me.

The  work will be done in following order:
* Fix the missed implementation of the CRS models.
* Apply the prevention error/exception handling to the other modules
* Implement the data vector/parameter checking

**Pull request merged**